XenAdmin/Dialogs/AttachUsbDialog.cs

```diff
@@ -56,16 +56,37 @@ private void BuildList()
             treeUsbList.BeginUpdate();
             try
             {
-                Host[] hosts = _vm.Connection.Cache.Hosts;
-                foreach (Host host in hosts)
+                VM.HA_Restart_Priority SelectedPriority = _vm.HARestartPriority();
+                Pool pool = Helpers.GetPool(_vm.Connection);
+                // Check if HA was enabled on pool and Restart priority was set on VM.
+                bool haEnabled = (pool != null &&
+                    pool.ha_enabled &&
+                    VM.HaPriorityIsRestart(_vm.Connection, SelectedPriority));
+                List<XenRef<Host>> possibleHostRefs = VM.get_possible_hosts(_vm.Connection.Session, _vm.opaque_ref);

```

Comment:
We shouldn't execute server calls on the UI thread. We can use a DelegateAsyncAction or can create a new action class.

Reply:
VM.HaPriorityIsRestart(IXenConnection connection, HA_Restart_Priority haRestartPriority) do not make any communication with server, but compare local fields. Parameter 'connection' is not used in fact. So it's safe to keep this code in UI. We called this method in some other UI pages/forms as well. Do you agree?

Reply:
Sorry, I think I misunderstood. You mean get_possible_hosts(), right?

Reply:
Yes